выполненное задание 8 по горутинам

FishSev opened the pull request:
1) пожалуйста загляни в файл Прочти.md там пояснения к решению 2) сама задачка понравилась, как всегда спасибо оргам (особенно Хамбару)

FishSev committed:
выполненное задание 8 по горутинам
1) пожалуйста загляни в файл Прочти.md там пояснения к решению
2) сама задачка понравилась, как всегда спасибо оргам (особенно Хамбару)

Tazya reviewed (commented):
Небольшие замечания по поводу функций.
writeCsvFile_plusChangeVal
readCsvFile_plusSort

Сразу по названию бросается в глаза, что у функции много полномочий.
В идеале она должна только записывать данные, принимая набор данных и путь.
А сортировка должна происходить где-то в другом месте.

Аналогчино и с чтением.

Хорошо раделять работу с внешним миром от вычислений.
Так снижается вероятность ошибок, появляется возможность написать юнит-тесты на сортировку, упрощается поиск багов и снижается нагрузка на читающего код.

Tazya reviewed (commented):
Сейчас по работе горутин ещё будет комментарий

Tazya commented on get_combination.go:
```diff
@@ -0,0 +1,17 @@
+	for i := 0; i <= 100; i++ {
+		wg.Add(1)
+		go pipeline.Pipeline(strconv.Itoa(i), &wg)
+		wg.Wait()
```
В таком варианте мы не получим профита по скорости от горутин.
Из-за того, что `wg.Wait()` приостанавливает цикл и ждёт пока горутина выполнит работу и только потом запускается следующая.
В итоге производительность и логика такая-же как при простом последовательном исполнении.

**Пошагово сейчас происходит так:**
```
1. Запуск цикла с i = 0;
2. wg.Add(1)
3. Обработка файла 0
4. Ожидание выполнения wg.Done()
5. Новая итерация цикла с i = 1;
6. ...
```

**Наша цель запустить горутины конкурентно, то есть добиться такого процесса:**
```
1. Запуск цикла
2. Запуск обработки файла 0 (не факт что первым по времени будет обработан dat0.csv :))
3. Запуск обработки файла 1
4. Запуск обработки файла 2
...
Ожидание выполнения всех wg.Done()
Завершение
```

Результат на моей системе такой.
1. Этот вариант - 260-300ms
2. Без горутин - 260-280ms
3. Исправленый wg - 100-120ms

FishSev committed:
переделал порядок цикла по горутинам

FishSev committed:
изменения
1) отработал по функциям, разделил обработку событий
writeCsvFile_plusChangeVal
readCsvFile_plusSort
2) убрал ненужные записи/чтения из файла, чем сократил время выполнения
   программы

Tazya reviewed (approved):
Approved! Круто получилось
